Increasing poling interval to 30 sec

daniyarchambylov opened the pull request.

daniyarchambylov committed:
Increasing poling interval to 30 sec

fgimenez reviewed (commented):
AIUI we should wait initially 30sec and then start polling with the schema we agree on (currently constant step == 1sec), but the step itself should not be 30sec, is that correct @rocky?

daniyarchambylov committed:
Merge branch 'master' into increase-polling-interval

daniyarchambylov committed:
Revert "Increasing poling interval to 30 sec"

This reverts commit 4c2b950120ed72383ca5cd5ca8ab9137c29ade93.

daniyarchambylov committed:
Delay initial polling request 30 secs

fgimenez reviewed (dismissed):
Nice, LGTM

rocky reviewed (changes requested):
Could we have a user option to set this initial delay value? 

It should be checked similar to the way we do in truffle-security

* is a number 
* is not less than the min(default value,`timeout`)

`timeout` refers to the value passed to `analyze()`, the maxiumum time the user is willing to wait on the request.

Thanks.

Commit:
Redo so we handle the initial delay more correctly.

* Allow a value to get passed in (if it is larger than 30 seconds)
* Test the response from analysis request for the "Finished" status
* Use min(defaultInitTimeout, 0.9 * timeout) as the first timeout value

Since we need to change things so that we return the first response
data (and not just the uuid as done previously), a lot of the unit
tests fail.

Daniyar please fix this up.

daniyarchambylov committed:
Fix tests for initial polling delay

daniyarchambylov committed:
Fix tests for lib/requester

birdofpreyru commented on index.js:
```diff
@@ -101,12 +112,37 @@ class Client {
+    // and then do the *only* subset of poller that needs to be done,
+    // given that we know this is finished. Instead, I think we'll
+    // make one more additional request of status only to find out
+    // that it is still finished.
```
Correct, if status is `Finished` at this point, there is no need to run entire poller logic, it is enough just to execute this piece directly: https://github.com/ConsenSys/armlet/blob/master/lib/poller.js#L71-L78

I'd say, split that piece into a separate `getIssues(..)` function inside `poller.js`, that will be called from `ping(..)`, and also exported and directly called here.

rocky replied:
Yeah, that is what I tried to do the other day and couldn't get it to work.

birdofpreyru replied:
Do you need help with it?

rocky replied:
YES! If you would do it, I'd be grateful.

birdofpreyru commented on index.js:
```diff
@@ -101,12 +112,37 @@ class Client {
+      let initialDelay = defaultInitialDelay
+      if (!isNaN(options.initialDelay) && options.initialDelay > initialDelay) {
+        initialDelay = options.initialDelay
+      }
```
Simplify:
```js
const initialDelay = Math.max(options.initialDelay || 0, defaultInitialDelay);
```

rocky replied:
Thanks! Will do.

birdofpreyru commented on index.js:
```diff
@@ -101,12 +112,37 @@ class Client {
+      // However, if the *total* timeout is small, we can use 90% of that
+      // instead, since we'll do just one or two polls anyway.
+
+      initialDelay = Math.min(initialDelay, 0.9 * timeout)
```
I'd say, no need for this. Even if caller ends up using `initialDelay > timeout`, then just wait for `timeout` and bail out. This way we keep the promise behind each argument in this corner case:
`timeout` specifies that the function will exit at most that time after the start,
`initialDelay` promises that no actuall API calls are done by function before this time from the start.

rocky replied:
After thinking about this a bit, I think you are correct.

To make it clear, here is what I had been thinking. In an IDE what you want is a short timeout say 1-5 seconds. That of course is too small to do an analysis. But that's okay, because the point was just to queue something up when there was indeed a change. 

In other words, setting timeout to say 1 second really means:  If the request was cached, return it. If not, just submit it and return.

Commit:
Incorporate Sergey's excellent suggestions

Commit:
No longer do the 90% thing - remove test

rocky committed:
Merge branch 'master' into increase-polling-interval

Commit:
I hate merge conflicts!

Fix one test, comment out another.

daniyarchambylov committed:
Fix commented test in index.js

daniyarchambylov committed:
Fix polling timeout

rocky committed:
Merge pull request #61 from ConsenSys/increase-polling-interval-fix-timeout

Fix polling timeout
